src/metabase/types.cljc

```diff
@@ -345,14 +345,32 @@
   [field :- SnakeCasedField]
   (field-is-type? :type/Temporal field))
 
+(def ^:private assignable-hierarchy
+  (make-hierarchy))
+
+(defn declare-assignable
+  "Declare that a value of type `x` assignable to a variable of type `y`."
+  [x y]
+  #?(:clj (alter-var-root #'assignable-hierarchy derive x y)
+     :cljs (set! assignable-hierarchy (derive assignable-hierarchy x y))))
+
+(declare-assignable :type/Integer :type/Decimal)
+
+(defn assignable?
+  "Is a value of type `x` assignable to a variable of type `y`?"
+  [x y]
+  (or (isa? assignable-hierarchy x y)
+      (boolean (some #(assignable? x %) (descendants y)))
+      (boolean (some #(assignable? % y) (parents x)))))
```

Comment:
Should `descendants` and `parents` here also be linked to the `assignable-hierarchy`? If we're deliberately using the larger `metabase.types` hierarchy, a comment saying so would be worthwhile.